cabal-testsuite/src/Test/Cabal/Prelude.hs

```diff
@@ -590,82 +592,77 @@ withRepo repo_dir m = do
-                | otherwise ->
-                  archiveTo srcPath (destPath <.> "tar.gz")
+        srcPath `archiveTo` sdistPath
+
+        -- When hackage-repo-tool extracts the cabal file from the tarball, it does carry
+        -- over the timestamp; so what ends up in the index is the time of this operation.
+        --
+        -- We extract the cabal file ourselves carrying over the modification time.
+        -- hackage-repo-tool would re-extract the cabal file is the sdist is newer, to
+        -- avoid this possibiliy we apply the same modification time to the sdist.
+        liftIO $ do
+          createDirectoryIfMissing True (takeDirectory idxPath)
+          copyFileWithMetadata (srcPath </> unPackageName pn <.> "cabal") idxPath
 
-    -- 3. Create keys and bootstrap repository
-    hackageRepoTool "create-keys" $ ["--keys", keysDir ]
-    hackageRepoTool "bootstrap" $ ["--keys", keysDir, "--repo", workDir]
+          ts <- System.Directory.getModificationTime (srcPath </> unPackageName pn <.> "cabal")
+          System.Directory.setModificationTime sdistPath ts
 
-    -- 4. Wire it up in .cabal/config
+    -- 5. Update repository
+    hackageRepoTool "update" ["--keys", testKeysDir env, "--repo", testRepoDir env]
+
+    -- 6. Wire it up in .cabal/config
     let package_cache = testCabalDir env </> "packages"
-    -- In the following we launch a python http server to serve the remote
-    -- repository. When the http server is ready we proceed with the tests.
-    -- NOTE 1: it's important that both the http server and cabal use the
-    -- same hostname ("localhost"), otherwise there could be a mismatch
-    -- (depending on the details of the host networking settings).
-    -- NOTE 2: here we use a fixed port (8000). This can cause problems in
-    -- case multiple tests are running concurrently or other another
-    -- process on the developer machine is using the same port.
-    liftIO $ do
-        appendFile (testUserCabalConfigFile env) $
-            unlines [ "repository repository.localhost"
-                    , "  url: http://localhost:8000/"
-                    , "  secure: True"
-                    , "  root-keys:"
-                    , "  key-threshold: 0"
-                    , "remote-repo-cache: " ++ package_cache ]
-        putStrLn $ testUserCabalConfigFile env
-        putStrLn =<< readFile (testUserCabalConfigFile env)
-
-        withAsync
-          (flip runReaderT env $ python3 ["-m", "http.server", "-d", workDir, "--bind", "localhost", "8000"])
-          (\_ -> do
-            -- wait for the python webserver to come up with a exponential
-            -- backoff starting from 50ms, up to a maximum wait of 60s
-            _ <- waitTcpVerbose putStrLn (limitRetriesByCumulativeDelay 60000000 $ exponentialBackoff 50000) "localhost" "8000"
-            runReaderT m (env { testHaveRepo = True }))
+    liftIO $ appendFile (testUserCabalConfigFile env)
```

Comment:
As an option, you could use `cabal user-config update`? But, perhaps, this doesn’t matter.

Reply:
uhm, TIL, how does that work? I have never used it before.

Reply:
Not sure what's the right way to call `cabal-install` in this part of the code, but from the command line you can do something like this to experiment:
```
❯ CABAL_DIR=. cabal update
❯ CABAL_DIR=. cabal user-config update --augment="repository test-local-repo
    url: file:some/dir/file
    secure: True
    root-keys: mykeys"
```

and observe `./config`. This seems to work for a `repository`, but trying to update `remote-repo-cache` doesn't do what is desired here: it updates the existing field from the default value `./packages` to whatever new value you supply. Unless I misunderstand what `appendFile` achieves: if it overrides the earlier value of `remote-repo-cache`, then maybe it's the same.

I see now that it's not your code: you just moved around what was there before, so I want to clarify that my remark about `user-config` shouldn't derail the PR.